### PR TITLE
Issue 709: Add Slow Bookkeeper Servers to Placement Policy for read ordering

### PR DESCRIPTION
**Pull request by philipsu522**
Descriptions of the changes in this PR:

Maintain a list of slow bookkeeper hosts, which are tried only after readonly bookkeeper hosts are tried. Bookkeeper hosts can be categorized as "slow" when a speculative timeout occurs on that bookkeeper host.

Master Issue:: 709

### CODE REVIEW

**Comment by philipsu522** [comment]
@jiazhai @ivankelly @eolivelli @sijie Sorry for the trouble, but I had to create a new pull request because I was accidentally rebasing between the wrong 2 branches locally, and there were a lot of conflicts on files I didn't change.

The changes are identical to the one in https://github.com/apache/bookkeeper/pull/776 (discarded), except a few refactors to incorporate latest changes in master.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 1]
```diff
@@ -789,35 +805,105 @@ protected BookieNode selectRandomFromRack(String netPath, Set<Node> excludeBooki
...
+        }
+        return reorderReadSequenceWithRegion(
+            ensemble, writeSet, writeSetWithRegion, bookiesHealthInfo, false, "", writeSet.size());
+    }
+
+    /**
+     * This function orders the read sequence with a given region. For region-unaware policies (e.g.
+     * RackAware), we pass in false for regionAware and an empty myRegion. When this happens, any
+     * remote list will stay empty. The ordering is as follows (the R* at the beginning of each list item
+     * is only present for region aware policies.
```
Comment (jiazhai) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java]: need a ")" ?

**Comment by sijie** [comment]
@eolivelli since you were involved in the original pull request #776, can you review this again?

File: .test-infra/jenkins/common_job_properties.groovy [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Contains functions that help build Jenkins projects. Functions typically set
+// common properties that are shared among all Jenkins projects.
+// Code in this directory should conform to the Groovy style guide.
+//  http://groovy-lang.org/style-guide.html
+class common_job_properties {
```
Comment (eolivelli) [inline .test-infra/jenkins/common_job_properties.groovy]: This is not part of this pr, maybe a rebase problem ?

Comment (sijie) [reply .test-infra/jenkins/common_job_properties.groovy]: @philipsu522 I think there is a problem on rebasing. can you merge latest master?
